kimchi/src/circuits/polynomials/chacha.rs

```diff
@@ -164,203 +166,53 @@ pub fn xor_table<F: Field>() -> Vec<Vec<F>> {
-            nyb(s[y], 4),
-            nyb(s[y], 5),
-            nyb(s[y], 6),
-            nyb(s[y], 7),
-        ]);
-
-        s[x] = xprime;
-        s[y] = yprime;
-
-        if k == 7 {
-            let lo = |t: u32, i: usize| f((t >> (4 * i)) & 1);
-            rows.push(vec![
-                f(yprime),
-                nyb(y_xor_xprime, 0),
-                nyb(y_xor_xprime, 1),
-                nyb(y_xor_xprime, 2),
-                nyb(y_xor_xprime, 3),
-                lo(y_xor_xprime, 0),
-                lo(y_xor_xprime, 1),
-                lo(y_xor_xprime, 2),
-                lo(y_xor_xprime, 3),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-            ]);
-            rows.push(vec![
-                F::zero(),
-                nyb(y_xor_xprime, 4),
-                nyb(y_xor_xprime, 5),
-                nyb(y_xor_xprime, 6),
-                nyb(y_xor_xprime, 7),
-                lo(y_xor_xprime, 4),
-                lo(y_xor_xprime, 5),
-                lo(y_xor_xprime, 6),
-                lo(y_xor_xprime, 7),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-                F::zero(),
-            ]);
-        }
-    };
-
-    let mut qr = |a, b, c, d| {
-        line(a, d, b, CHACHA20_ROTATIONS[0]);
-        line(c, b, d, CHACHA20_ROTATIONS[1]);
-        line(a, d, b, CHACHA20_ROTATIONS[2]);
-        line(c, b, d, CHACHA20_ROTATIONS[3]);
-    };
-    for _ in 0..10 {
-        for [a, b, c, d] in CHACHA20_QRS {
-            qr(a, b, c, d);
-        }
-    }
-
-    rows
-}
+/// Implementation details
+struct Impl<F>(PhantomData<F>);
 
-pub fn chacha20(mut s: Vec<u32>) -> Vec<u32> {
-    let mut line = |x, y, z, k| {
-        s[x] = u32::wrapping_add(s[x], s[z]);
-        s[y] ^= s[x];
-        let yy: u32 = s[y];
-        s[y] = yy.rotate_left(k);
-    };
-    let mut qr = |a, b, c, d| {
-        line(a, d, b, CHACHA20_ROTATIONS[0]);
-        line(c, b, d, CHACHA20_ROTATIONS[1]);
-        line(a, d, b, CHACHA20_ROTATIONS[2]);
-        line(c, b, d, CHACHA20_ROTATIONS[3]);
-    };
-    for _ in 0..10 {
-        for [a, b, c, d] in CHACHA20_QRS {
-            qr(a, b, c, d);
-        }
+impl<F> Impl<F>
+where
+    F: FftField,
+{
+    fn w(i: usize, r: CurrOrNext) -> E<F> {
+        E::<F>::cell(Column::Witness(i), r)
     }
-    s
-}
-
-// TODO: Assert somehow that we're using the right number of alphas.
-pub fn constraint<F: FftField>(alpha0: usize) -> E<F> {
-    let w = |i: usize, r: CurrOrNext| E::<F>::cell(Column::Witness(i), r);
 
-    // 8-nybble sequences that are laid out as 4 nybbles per row over the two row,
-    // like y^x' or x+z
-    let chunks_over_2_rows = |col_offset: usize| -> Vec<_> {
+    /// 8-nybble sequences that are laid out as 4 nybbles per row over the two row,
```

Comment:
Is this a doc-comment on a private function? :thinking:

Reply:
yeah, you can still export private function comments if you want with a flag, so that's useful :o